chore(docs): Add Snipcart reference guide (#22628)
gatsbyjs/gatsby: docs/docs/adding-a-shopping-cart-with-snipcart.md

  <button
    class="snipcart-add-item"
    data-item-id="silver-stacking-ring"
    data-item-price="19.99"
    data-item-url="/"
    data-item-name="Silver Stacking Ring"
    <!--highlight-start-->
    data-item-custom1-name="Size"
    data-item-custom1-options="6|6.5|7|7.5|8|8.5|9"
    <!-- highlight-end -->
    >
    Add to cart
  </button>
</section>
```

You can add multiple custom fields by incrementing the index of the `data-item-custom` attribute. Perhaps you want customers to have the ability to mark each item in their cart as a gift.

```jsx
<section>
  <h2>Silver Stacking Ring</h2>
  <p>$19.99</p>
  <p>Wear one or seventeen! These rings are fun to mix and match.</p>
  <button
    class="snipcart-add-item"
    data-item-id="silver-stacking-ring"
    data-item-price="19.99"
    data-item-url="/"
    data-item-name="Silver Stacking Ring"
    data-item-custom1-name="Size"
    data-item-custom1-options="6|6.5|7|7.5|8|8.5|9"
    <!--highlight-start-->
    data-item-custom2-name="This is a gift" data-item-custom2-type="checkbox"
   >
    <!-- highlight-end -->
    Add to cart
  </button>
</section>
```

### Selling digital products

> Snipcart enables the sale of digital goods. To sell a file, you'll need to upload it to your Snipcart dashboard and then add the resulting guid as the value of the `data-item-file-guid` attribute to your product's markup. You can specify a file access expiry in days and a maximum number of downloads per order from the dashboard.

Author: ```suggestion
Snipcart enables the sale of digital goods such as e-books, photography, and other artwork. To sell a file, you'll need to upload it to your Snipcart dashboard and then add the resulting guid as the value of the `data-item-file-guid` attribute to your product's markup. You can specify a file access expiry in days and a maximum number of downloads per order from the dashboard.
```

Reviewer: Hadn't seen this edit. I'm good with this or the above, whichever you prefer.

---
title: Adding a Shopping Cart with Snipcart
---

> Snipcart is a shopping cart solution designed to be dropped neatly into any web project. Including it in your project allows any HTML you write to instantly become a "buy button" with the addition of several Snipcart-specific attributes.

Reviewer: ```suggestion
Snipcart is a shopping cart solution designed to drop neatly into any web project. Including it in your project allows any HTML you write to instantly become a "buy button" with the addition of several Snipcart-specific attributes.
```

> Combine it with a source of products (like a CMS or an e-commerce platform like [Etsy](https://www.etsy.com/)) and a payment processor (like [Stripe](https://www.stripe.com/)) to build a complete e-commerce experience for your customers.

Reviewer: ```suggestion
Combine it with a source of products (like a CMS or an e-commerce platform such as [Etsy](https://www.etsy.com/)) and a payment processor (like [Stripe](https://www.stripe.com/)) to build a complete e-commerce experience for your customers.
```

## Prerequisites

To get started, you'll need to have the following set up:

- A Gatsby site with [`gatsby-plugin-snipcart`](/packages/gatsby-plugin-snipcart/) installed
- A [Snipcart](https://snipcart.com/) account
- A Snipcart test API key
- A list of products to sell

> - a [Snipcart](https://snipcart.com/) account

Reviewer: ```suggestion
```

Author: 🤦🏽‍♀️

Reviewer: Lol, happens to all of us

> - your Snipcart test API key

Reviewer: ```suggestion
```

> - products to sell

Reviewer: ```suggestion
```

Author: I... am not sure how I even did this.

## Defining Products

Adding products with Snipcart involves writing HTML representing your product and adding a set of attributes to that HTML. You might write something similar to the following code block for each item in your catalog. This code could be part of a page like `index.js` or anywhere else you list a product.

```jsx
<section>
  <h2>Silver Stacking Ring</h2>
  <p>$19.99</p>
  <p>Wear one or seventeen! These rings are fun to mix and match.</p>
</section>
```

Including this information allows a visitor to see what you have for sale, but they can't do anything with that information quite yet. You'll need a way for customers to add individual items to their cart. Try adding a button with the following attributes.

```jsx
<section>
  <h2>Silver Stacking Ring</h2>
  <p>$19.99</p>
  <p>Wear one or seventeen! These rings are fun to mix and match.</p>
  <!-- highlight-start -->
  <button
    class="snipcart-add-item"
    data-item-id="silver-stacking-ring"
    data-item-price="19.99"
    data-item-url="/"
    data-item-name="Silver Stacking Ring"
  >
    Add to cart
  </button>
  <!-- highlight-end -->
</section>
```

> Snipcart uses these attributes (`data-item-*`) to figure out what your customer is trying to buy and how much to charge them. The ID, price, URL, and name attributes are all required but there are several others that you can add to enhance the shopping cart.

Reviewer: ```suggestion
Snipcart uses these attributes (`data-item-*`) to figure out what your customer is trying to buy and how much to charge them. The ID, price, URL, and name attributes are all required but there are several other attributes that you can add to enhance the shopping cart.
```

  <button
    class="snipcart-add-item"
    data-item-id="silver-stacking-ring"
    data-item-price="19.99"
    data-item-url="/"
    data-item-name="Silver Stacking Ring"
    <!--highlight-start-->
    data-item-custom1-name="Size"
    data-item-custom1-options="6|6.5|7|7.5|8|8.5|9"
    <!-- highlight-end -->
    >
    Add to cart
  </button>
</section>
```

You can add multiple custom fields by incrementing the index of the `data-item-custom` attribute. Perhaps you want customers to have the ability to mark each item in their cart as a gift.

```jsx
<section>
  <h2>Silver Stacking Ring</h2>
  <p>$19.99</p>
  <p>Wear one or seventeen! These rings are fun to mix and match.</p>
  <button
    class="snipcart-add-item"
    data-item-id="silver-stacking-ring"
    data-item-price="19.99"
    data-item-url="/"
    data-item-name="Silver Stacking Ring"
    data-item-custom1-name="Size"
    data-item-custom1-options="6|6.5|7|7.5|8|8.5|9"
    <!--highlight-start-->
    data-item-custom2-name="This is a gift" data-item-custom2-type="checkbox"
   >
    <!-- highlight-end -->
    Add to cart
  </button>
</section>
```

### Selling digital products

> Snipcart enables the sale of digital goods. To sell a file, you'll need to upload it to your Snipcart dashboard and then add the resulting guid as the value of the `data-item-file-guid` attribute to your product's markup. You can specify a file access expiry in days and a maximum number of downloads per order from the dashboard.

Reviewer: Not sure I understand "file" in this context?

Author: As in a file you would like people to pay to download. I felt like "digital good" was a bit vague but that's how they describe it in the Snipcart docs. Open to suggestions.

Reviewer: Ahh, ok, so the heading context was totally lost on me. And I just assumed digital goods meant buying stuff via the web rather than the product itself being downloadable. Perhaps that's what needs to be clarified? So `Snipcart enables the sale of digital goods that consumers download directly to their computer.`? It's redundant but helps hammer home the shift from online shopping to online goods.

---
title: Adding a Shopping Cart with Snipcart
---

Snipcart is a shopping cart solution designed to be dropped neatly into any web project. Including it in your project allows any HTML you write to instantly become a "buy button" with the addition of several Snipcart-specific attributes.

Combine it with a source of products (like a CMS or an e-commerce platform like [Etsy](https://www.etsy.com/)) and a payment processer (like [Stripe](https://www.stripe.com/)) to build a complete e-commerce experience for your customers.

## Prerequisites

To get started, you'll need

- a Gatsby site with [`gatsby-plugin-snipcart`](/packages/gatsby-plugin-snipcart/) installed
- a [Snipcart](https://snipcart.com/) account
- your Snipcart test API key
- products to sell

Installing the plugin adds Snipcart's shopping cart JavaScript for you, so you can get right to building your e-commerce site. It's okay if you're not sure what you'd like to sell quite yet. Using sample products is fine to begin with!

## Defining Products

> Adding products with Snipcart involves writing HTML representing your product and adding a set of attributes to that HTML. You might write something similar to the following code block for each item in your catalogue.

Author: Switched to JSX! I didn't add a specific file because this would depend on where you listed your products. Maybe something a little less prescriptive like "wherever you list items for sale..."

To learn more about defining products, see the [Snipcart documentation](https://docs.snipcart.com/v3/setup/products).

### Adding product variants

Snipcart refers to variations like size and color as "product options" or "custom fields". You can add these custom fields to your products to allow customers to refine their orders.

Building on the stacking ring example, suppose that you wanted to give your customer a choice between available sizes. You would do this by adding a custom field with a name and options.

```html
<section>
  <h2>Silver Stacking Ring</h2>
  <p>$19.99</p>
  <p>Wear one or seventeen! These rings are fun to mix and match.</p>
  <button
    class="snipcart-add-item"
    data-item-id="silver-stacking-ring"
    data-item-price="19.99"
    data-item-url="/"
    data-item-name="Silver Stacking Ring"
    <!--
    highlight-start
    --
  >
    data-item-custom1-name="Size"
    data-item-custom1-options="6|6.5|7|7.5|8|8.5|9">
    <!-- highlight-end -->
    Add to cart
  </button>

> </section>

Reviewer 2: ```suggestion
  <button
    class="snipcart-add-item"
    data-item-id="silver-stacking-ring"
    data-item-price="19.99"
    data-item-url="/"
    data-item-name="Silver Stacking Ring"
    <!--highlight-start-->
    data-item-custom1-name="Size"
    data-item-custom1-options="6|6.5|7|7.5|8|8.5|9"
    <!-- highlight-end -->
    >
    Add to cart
  </button>
</section>
```

Author: I'm not sure how that even happened! Prettier? Good catch!

```

You can add multiple custom fields by incrementing the index of the `data-item-custom` attribute. Perhaps you want customers to have the ability to mark each item in their cart as a gift.

```html
<section>
  <h2>Silver Stacking Ring</h2>
  <p>$19.99</p>
  <p>Wear one or seventeen! These rings are fun to mix and match.</p>
  <button
    class="snipcart-add-item"
    data-item-id="silver-stacking-ring"
    data-item-price="19.99"
    data-item-url="/"
    data-item-name="Silver Stacking Ring"
    data-item-custom1-name="Size"
    data-item-custom1-options="6|6.5|7|7.5|8|8.5|9"
    <!--
    highlight-start
    --
  >
    data-item-custom2-name="This is a gift" data-item-custom2-type="checkbox">
    <!-- highlight-end -->
    Add to cart
  </button>

> </section>

Reviewer 2: ```suggestion
<section>
  <h2>Silver Stacking Ring</h2>
  <p>$19.99</p>
  <p>Wear one or seventeen! These rings are fun to mix and match.</p>
  <button
    class="snipcart-add-item"
    data-item-id="silver-stacking-ring"
    data-item-price="19.99"
    data-item-url="/"
    data-item-name="Silver Stacking Ring"
    data-item-custom1-name="Size"
    data-item-custom1-options="6|6.5|7|7.5|8|8.5|9"
    <!--highlight-start-->
    data-item-custom2-name="This is a gift" data-item-custom2-type="checkbox"
   >
    <!-- highlight-end -->
    Add to cart
  </button>
</section>
```

```

### Selling digital products

Snipcart enables the sale of digital goods. To sell a file, you'll need to upload it to your Snipcart dashboard and then add the resulting guid as the value of the `data-item-file-guid` attribute to your product's markup. You can specify a file access expiry in days and a maximum number of downloads per order from the dashboard.

```html
<section>
  <h2>Silver Stacking Ring</h2>
  <p>$19.99</p>
  <p>Wear one or seventeen! These rings are fun to mix and match.</p>
  <button
    class="snipcart-add-item"
    data-item-id="silver-stacking-ring"
    data-item-price="19.99"
    data-item-url="/"
    data-item-name="Silver Stacking Ring"
    <!--
    highlight-start
    --
  >
    data-item-file-guid="your-digital-product-guid">
    <!-- highlight-end -->
    Add to cart
  </button>

> </section>

Reviewer 2: ```suggestion
<section>
  <h2>Silver Stacking Ring</h2>
  <p>$19.99</p>
  <p>Wear one or seventeen! These rings are fun to mix and match.</p>
  <button
    class="snipcart-add-item"
    data-item-id="silver-stacking-ring"
    data-item-price="19.99"
    data-item-url="/"
    data-item-name="Silver Stacking Ring"
    <!--highlight-start-->
    data-item-file-guid="your-digital-product-guid">
    <!-- highlight-end -->
    Add to cart
  </button>
</section>
```

```

## Customizing the cart

Using Snipcart allows you to retain nearly complete control over your customers' experience on your e-commerce site. You can configure and customize the cart behavior as well as the product options. Look through your [Snipcart account settings](https://app.snipcart.com/dashboard/account/settings) to change things like currency, shipping options, and email templates.

### Preventing automatic popups

> By default, the shopping cart will pop up every time a customer adds a product. To prevent this bahavior, set the value of `autopop` to `false` in your `gatsby-config.js` file.

Author: ```suggestion
By default, the shopping cart will pop up every time a customer adds a product. To prevent this behavior, set the value of `autopop` to `false` in your `gatsby-config.js` file.
```

```js:title=gatsby-config.js
{
  resolve: 'gatsby-plugin-snipcart',
  options: {
    apiKey: 'your-api-key',
    autopop: false // highlight-line
  }
},
```

If you choose to prevent this popup, you'll need to give your customers some other way to access their shopping carts. Create a "show cart" button by giving a `button` element a class of `snipcart-checkout`.

```jsx
<button className="snipcart-checkout">View Cart</button>
```

### Styling the cart

> You can override most aspects of the shopping cart, including the CSS. Try inspecting the element you'd like to customize and using your browser's developer tools to find the correct Snipcart class to override.

Reviewer 2: Is the CSS file created and then Imported into the `gatsby-browser.js` file so it's discoverable?

Author: No, just linked by the plugin if I recall correctly. I literally inspected the specific elements I wanted to update or read the docs to find the correct classes for things.

Author: I feel like this is covered by the theme reference link in this section. I don't know of another/better resource for this kind of modification.

### Selling digital products

> Snipcart enables the sale of digital goods such as e-books, photography, and other artwork. To sell a file, you'll need to upload it to your Snipcart dashboard and then add the resulting guid as the value of the `data-item-file-guid` attribute to your product's markup. You can specify a file access expiry in days and a maximum number of downloads per order from the dashboard.

Reviewer: ```suggestion
Snipcart enables the sale of digital goods such as e-books, photography, and other artwork. To sell a file you intend for download, you'll need to upload it to your Snipcart dashboard and then add the resulting guid as the value of the `data-item-file-guid` attribute to your product's markup. You can specify a file access expiry in days and a maximum number of downloads per order from the dashboard.
```
One last idea to make this as clear as possible. Feel free to disagree.
